hbase-procedure/src/main/java/org/apache/hadoop/hbase/procedure2/ProcedureExecutor.java

```diff
@@ -1971,27 +2002,7 @@ public void run() {
             continue;
           }
           this.activeProcedure = proc;
-          int activeCount = activeExecutorCount.incrementAndGet();
-          int runningCount = store.setRunningProcedureCount(activeCount);
-          LOG.trace("Execute pid={} runningCount={}, activeCount={}", proc.getProcId(),
-            runningCount, activeCount);
-          executionStartTime.set(EnvironmentEdgeManager.currentTime());
-          IdLock.Entry lockEntry = procExecutionLock.getLockEntry(proc.getProcId());
-          try {
-            executeProcedure(proc);
-          } catch (AssertionError e) {
-            LOG.info("ASSERT pid=" + proc.getProcId(), e);
-            throw e;
-          } finally {
-            procExecutionLock.releaseLockEntry(lockEntry);
-            activeCount = activeExecutorCount.decrementAndGet();
-            runningCount = store.setRunningProcedureCount(activeCount);
-            LOG.trace("Halt pid={} runningCount={}, activeCount={}", proc.getProcId(),
-              runningCount, activeCount);
-            this.activeProcedure = null;
-            lastUpdate = EnvironmentEdgeManager.currentTime();
-            executionStartTime.set(Long.MAX_VALUE);
-          }
+          lastUpdate = TraceUtil.trace(this::runProcedure, new ProcedureSpanBuilder(proc));
```

Comment:
In a procedure's whole life time, it could reach here multiple times. For example, if it schedules some sub procedures, it will suspend itself for a while until all the sub procedures are finished, it will recover the execution. So I think if we want to trace procedure execution, we need to discuss more, and better to have a design doc first. It is not very straight forward, as how to define the relationship between parent procedure and sub procedure?

Reply:
This implementation is very naive. There is no relationship between parent and child procedure. There is no concept of tying together multiple suspend/resume executions of the same procedure ID. It's merely a way to recognize that all the resulting RPCs are tied back to a procedure by ID. We can explore those additional features, and I agree that effort should include a lot more thought.

Reply:
I can update the JIRA title to make it sound less thorough in its scope, if that helps at all.

Reply:
Something like 'Trace a one-shot execution of a procedure' seems more accurate :)